src/sha1.rs

```diff
@@ -22,118 +22,368 @@
+///
+/// A SHA-1 implementation derived from Paul E. Jones's reference
+/// implementation, which is written for clarity, not speed. At some
+/// point this will want to be rewritten.
+#[inline]
+pub fn sha1_digest_block_u32_safe(state: &mut [u32/*; 5*/], block: &[u32/*; 16*/]) {
+    let mut w = [0u32; 80];
+
+    let mut a = state[0];
+    let mut b = state[1];
+    let mut c = state[2];
+    let mut d = state[3];
+    let mut e = state[4];
+    let mut temp: u32;
 
     // Initialize the first 16 words of the vector w
-    read_u32v_be(&mut w[0..16], data);
+    for t in 0..16 {
+        w[t] = block[t].to_be();
+    }
 
     // Initialize the rest of vector w
     let mut t = 16; // loop counter
     while t < 80 {
         let val = w[t - 3] ^ w[t - 8] ^ w[t - 14] ^ w[t - 16];
-        w[t] = circular_shift(1, val);
+        w[t] = val.rotate_left(1);
         t += 1;
     }
-    let mut a = h[0];
-    let mut b = h[1];
-    let mut c = h[2];
-    let mut d = h[3];
-    let mut e = h[4];
-    let mut temp: u32;
     t = 0;
     while t < 20 {
-        temp = circular_shift(5, a) + (b & c | !b & d) + e + w[t] + K0;
+        temp = a.rotate_left(5) + (b & c | !b & d) + e + w[t] + K0;
         e = d;
         d = c;
-        c = circular_shift(30, b);
+        c = b.rotate_left(30);
         b = a;
         a = temp;
         t += 1;
     }
     while t < 40 {
-        temp = circular_shift(5, a) + (b ^ c ^ d) + e + w[t] + K1;
+        temp = a.rotate_left(5) + (b ^ c ^ d) + e + w[t] + K1;
         e = d;
         d = c;
-        c = circular_shift(30, b);
+        c = b.rotate_left(30);
         b = a;
         a = temp;
         t += 1;
     }
     while t < 60 {
-        temp =
-            circular_shift(5, a) + (b & c | b & d | c & d) + e + w[t] +
-                K2;
+        temp = a.rotate_left(5) + (b & c | b & d | c & d) + e + w[t] + K2;
         e = d;
         d = c;
-        c = circular_shift(30, b);
+        c = b.rotate_left(30);
         b = a;
         a = temp;
         t += 1;
     }
     while t < 80 {
-        temp = circular_shift(5, a) + (b ^ c ^ d) + e + w[t] + K3;
+        temp = a.rotate_left(5) + (b ^ c ^ d) + e + w[t] + K3;
         e = d;
         d = c;
-        c = circular_shift(30, b);
+        c = b.rotate_left(30);
         b = a;
         a = temp;
         t += 1;
     }
-    h[0] += a;
-    h[1] += b;
-    h[2] += c;
-    h[3] += d;
-    h[4] += e;
+
+    state[0] += a;
+    state[1] += b;
+    state[2] += c;
+    state[3] += d;
+    state[4] += e;
+}
+
+/// Process a block with the SHA-1 algorithm.
+///
+/// Chooses an implementation based on architecture,
+/// and whether or not the architecture supports SHA
+/// instruction set extensions.
+pub fn sha1_digest_block(state: &mut [u32/*; 5*/], bytes: &[u8/*; 64*/]) {
+    assert_eq!(state.len(), STATE_LEN);
+    assert_eq!(bytes.len(), BLOCK_LEN*4);
+    let (words, _): (&[u32; 16], usize) = unsafe {
+        ::std::mem::transmute(bytes)
```

Comment:
This transmute can cause issues on architectures that require alignment of their types. ARM is the primary example that I know of. The benchmarks don't seem to hit this issue. However, the tests do which cause the test program to crash with a SIGBUS. Can you think of a way to avoid the transmute or to only do the transmute if it the alignment is right? Rust does provide an align_of() function.